samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,253 @@
+
+    def _remove_unnecessary_fields(self, template_dict: Dict, linked_resources: List[str] = []) -> List[str]:
+        """
+        Fields skipped during template comparison because sync --code can handle the difference:
+        * CodeUri or ImageUri property of AWS::Serverless::Function
+        * ImageUri, S3Bucket, S3Key, S3ObjectVersion fields in Code property of AWS::Lambda::Function
+        * ContentUri property of AWS::Serverless::LayerVersion
+        * Content property of AWS::Lambda::LayerVersion
+        * DefinitionBody property of AWS::Serverless::Api
+        * BodyS3Location property of AWS::ApiGateway::RestApi
+        * DefinitionUri property of AWS::Serverless::HttpApi
+        * BodyS3Location property of AWS::ApiGatewayV2::Api
+        * DefinitionUri property of AWS::Serverless::StateMachine
+        * DefinitionS3Location property of AWS::StepFunctions::StateMachine
+        * SamResourceId in Metadata property of all resources
+
+        Parameters
+        ----------
+        template_dict: Dict
+            The unprocessed template dictionary
+        linked_resources: List[str]
+            The corresponding resources in the other template that got processed
+
+        Returns
+        -------
+        List[str]
+            The list of resource IDs that got changed during sanitization
+        """
+
+        resources = template_dict.get("Resources", {})
+        processed_resources: Set[str] = set()
+
+        for resource_logical_id in resources:
+            resource_dict = resources.get(resource_logical_id, {})
+            resource_type = resource_dict.get("Type")
+
+            if resource_type in SYNCABLE_RESOURCES:
+                processed_resources = self._remove_resource_field(
+                    resource_logical_id,
+                    resource_type,
+                    resource_dict,
+                    processed_resources,
+                    linked_resources,
+                )
+
+            # Remove SamResourceId metadata since this metadata does not affect any cloud behaviour
+            resource_dict.get("Metadata", {}).pop("SamResourceId", None)
+            if not resource_dict.get("Metadata"):
+                resource_dict.pop("Metadata", None)
+            LOG.debug(f"Sanitizing the Metadata for resource {resource_logical_id}")
+
+        return sorted(list(processed_resources))
+
+    def _remove_resource_field(
+        self,
+        resource_logical_id: str,
+        resource_type: str,
+        resource_dict: Dict,
+        processed_resources: set,
+        linked_resources: List[str] = [],
+    ) -> set:
+        """
+        Helper method to process resource dict
+
+        Parameters
+        ----------
+        resource_logical_id: str
+            Logical ID of the resource
+        resource_type: str
+            Resource type
+        resource_dict: Dict
+            The resource level dict containing Properties field
+        processed_resources: set
```

Comment:
Curious why we pass in processed_resources parameter for this function? Does it make sense to update the processed_resources set in `_remove_unnecessary_fields` function and change the return type of `_remove_resource_field` function to be logical_id of the removed resource? This way processed_resources variable doesn't need to be passed to `_remove_resource_field` function.

Reply:
Good call out, we can have one less parameter since we will only add one resource ID each time, so we only need to return a string